diff --git a/panels.module b/panels.module
index 7c33a7d..c496735 100644
--- a/panels.module
+++ b/panels.module
@@ -1708,3 +1708,21 @@ function panels_get_path($file, $base_path = FALSE, $module = 'panels') {
   $output = $base_path ? base_path() : '';
   return $output . drupal_get_path('module', $module) . '/' . $file;
 }
+
+/**
+ * Remove default sidebar related body classes and provide own css classes
+ */
+function panels_preprocess_html(&$vars) {
+  $panel_body_css = &drupal_static('panel_body_css');
+  if (!empty($panel_body_css['body_classes_to_remove'])) {
+    $classes_to_remove = explode(' ', $panel_body_css['body_classes_to_remove']);
+    foreach ($vars['classes_array'] as $key => $css_class) {
+      if (in_array($css_class, $classes_to_remove)) {
+        unset($vars['classes_array'][$key]);
+      }
+    }
+  }
+  if (!empty($panel_body_css['body_classes_to_add'])) {
+    $vars['classes_array'][] = check_plain($panel_body_css['body_classes_to_add']);
+  }
+}
diff --git a/plugins/task_handlers/panel_context.inc b/plugins/task_handlers/panel_context.inc
index 86512f6..4f2da19 100644
--- a/plugins/task_handlers/panel_context.inc
+++ b/plugins/task_handlers/panel_context.inc
@@ -185,6 +185,8 @@ $plugin = array(
     'title' => t('Panel'),
     'no_blocks' => FALSE,
     'pipeline' => 'standard',
+    'body_classes_to_remove' => '',
+    'body_classes_to_add' => '',
     'css_id' => '',
     'css' => '',
     'contexts' => array(),
@@ -283,6 +285,16 @@ function panels_panel_context_render($handler, $base_contexts, $args, $test = TR

   $renderer = panels_get_renderer($handler->conf['pipeline'], $display);

+  // Remove and add body element classes
+  $panel_body_css = &drupal_static('panel_body_css');
+
+  if (isset($handler->conf['body_classes_to_remove']) && !isset($panel_body_css['body_classes_to_remove'])) {
+    $panel_body_css['body_classes_to_remove'] = $handler->conf['body_classes_to_remove'];
+  }
+  if (isset($handler->conf['body_classes_to_add']) && !isset($panel_body_css['body_classes_to_add'])) {
+    $panel_body_css['body_classes_to_add'] = $handler->conf['body_classes_to_add'];
+  }
+
   $info = array(
     'content' => panels_render_display($display, $renderer),
     'no_blocks' => !empty($handler->conf['no_blocks']),
@@ -714,6 +726,22 @@ function panels_panel_context_edit_settings($form, &$form_state) {
     '#description' => t('Check this to have the page disable all regions displayed in the theme. Note that some themes support this setting better than others. If in doubt, try with stock themes to see.'),
   );

+  $form['conf']['body_classes_to_remove'] = array(
+    '#type' => 'textfield',
+    '#size' => 128,
+    '#default_value' => $conf['body_classes_to_remove'],
+    '#title' => t('Remove body CSS classes'),
+    '#description' => t('The CSS classes to remove from the body element of this page. Separated by a space. For example: no-sidebars one-sidebar sidebar-first sidebar-second two-sidebars.'),
+  );
+
+  $form['conf']['body_classes_to_add'] = array(
+    '#type' => 'textfield',
+    '#size' => 128,
+    '#default_value' => $conf['body_classes_to_add'],
+    '#title' => t('Add body CSS classes'),
+    '#description' => t('The CSS classes to add to the body element of this page. Separated by a space. For example: no-sidebars one-sidebar sidebar-first sidebar-second two-sidebars.'),
+  );
+
   ctools_include('plugins', 'panels');
   $pipelines = panels_get_renderer_pipelines();

@@ -769,6 +797,8 @@ function panels_panel_context_edit_settings($form, &$form_state) {
  */
 function panels_panel_context_edit_settings_submit(&$form, &$form_state) {
   $form_state['handler']->conf['no_blocks'] = $form_state['values']['no_blocks'];
+  $form_state['handler']->conf['body_classes_to_remove'] = $form_state['values']['body_classes_to_remove'];
+  $form_state['handler']->conf['body_classes_to_add'] = $form_state['values']['body_classes_to_add'];
   $form_state['handler']->conf['pipeline'] = $form_state['values']['pipeline'];
   $form_state['handler']->conf['css_id'] = $form_state['values']['css_id'];
   $form_state['handler']->conf['css'] = $form_state['values']['css'];
 